Source/CarthageKit/DependencyCrawler.swift

```diff
@@ -0,0 +1,226 @@
+					throw DependencyCrawlerError.versionRetrievalFailure(message: "Could not collect versions for dependency: \(dependency) and versionSpeficier: \(versionSpecifier)")
+				}
+				cachedVersionSets[dependency] = pinnedVersions
+
+				let storedDependency = self.dependencyMappings?[dependency] ?? dependency
+				try store.storePinnedVersions(pinnedVersions, for: storedDependency, gitReference: gitReference)
+
+				versionSet = pinnedVersions
+			}
+
+			let filteredVersionSet = versionSet.filter { pinnedVersion -> Bool in
+				versionSpecifier.isSatisfied(by: pinnedVersion)
+			}
+
+			eventPublisher.send(value:
+				.foundVersions(versions: filteredVersionSet, dependency: dependency, versionSpecifier: versionSpecifier)
+			)
+
+			return filteredVersionSet
+		} catch let error {
+
+			eventPublisher.send(value:
+				.failedRetrievingVersions(error: error, dependency: dependency, versionSpeficier: versionSpecifier)
+			)
+
+			if ignoreErrors {
+				return [PinnedVersion]()
+			} else {
+				throw error
+			}
+		}
+	}
+
+	private func findDependencies(for dependency: Dependency, version: PinnedVersion) throws -> [(Dependency, VersionSpecifier)] {
+		do {
+			guard let transitiveDependencies: [(Dependency, VersionSpecifier)] = try dependenciesForDependency(dependency, version).collect().first()?.dematerialize() else {
+				throw DependencyCrawlerError.dependencyRetrievalFailure(message: "Could not find transitive dependencies for dependency: \(dependency), version: \(version)")
+			}
+
+			let storedDependency = self.dependencyMappings?[dependency] ?? dependency
+			let storedTransitiveDependencies = transitiveDependencies.map { transitiveDependency, versionSpecifier -> (Dependency, VersionSpecifier) in
+				let storedTransitiveDependency = self.dependencyMappings?[transitiveDependency] ?? transitiveDependency
+				return (storedTransitiveDependency, versionSpecifier)
+			}
+			try store.storeTransitiveDependencies(storedTransitiveDependencies, for: storedDependency, version: version)
+
+			eventPublisher.send(value:
+				.foundTransitiveDependencies(transitiveDependencies: transitiveDependencies, dependency: dependency, version: version)
+			)
+
+			return transitiveDependencies
+		} catch let error {
+
+			eventPublisher.send(value:
+				.failedRetrievingTransitiveDependencies(error: error, dependency: dependency, version: version)
+			)
+
+			if ignoreErrors {
+				return [(Dependency, VersionSpecifier)]()
+			} else {
+				throw error
+			}
+		}
+	}
+}
+
+private struct PinnedDependency: Hashable {
+	public let dependency: Dependency
+	public let pinnedVersion: PinnedVersion
+	private let hash: Int
+
+	init(dependency: Dependency, pinnedVersion: PinnedVersion) {
+		self.dependency = dependency
+		self.pinnedVersion = pinnedVersion
+		// Pre-compute hash for efficiency
+		self.hash = 37 &* dependency.hashValue &+ pinnedVersion.hashValue
```

Comment:
I don't think we should be doing this. Did you actually instrument and find this to be a source of slowness?

Reply:
As part of the resolver improvement MR I indeed found that hash calculation is one of the bottlenecks for the performance issues.

Reply:
Removed it to rely on auto-conformance of Hashable. For this particular case performance is not that critical.